### DISCUSSION TITLE
QueryRaw variables and type casting?

### DISCUSSION THREAD

**Discussion by leo-petrucci** [prisma/prisma#9190] [Q&A] [answered]
I'm having trouble with a query, I think there might be something not working typecasts while using variables. Using Postgres.

My query is:
```
      return await prisma.$queryRaw`
        SELECT
          d::date as key
        FROM generate_series(
          ${dayjs(fromDate).toISOString()},
          ${dayjs(toDate).toISOString()},
          '1 day'
        ) AS gs(d)
        GROUP BY
          gs.d
        ORDER BY 
          d::date ASC
      `;
```
This should return an array of dates between `fromDate` and `toDate`.

When I try to run I get the error:
```
            '  Raw query failed. Code: `42725`. Message: `db error: ERROR: function generate_series(unknown, unknown, unknown) is not unique\n' +
            'HINT: Could not choose a best candidate function. You might need to add explicit type casts.`',
```

So I need to typecast, okay:
```
          ${dayjs(fromDate).toISOString()}::timestamp,
          ${dayjs(toDate).toISOString()}::timestamp,
```
Returns:
```
            '  Raw query failed. Code: `22P03`. Message: `db error: ERROR: incorrect binary data format in bind parameter 1`',
```

```
          timestamp ${dayjs(fromDate).toISOString()},
          timestamp ${dayjs(toDate).toISOString()},
```
Returns:
```
            '  Raw query failed. Code: `42601`. Message: `db error: ERROR: syntax error at or near "$1"`',
```

```
          CAST(${dayjs(fromDate).toISOString()} as timestamp),
          CAST(${dayjs(toDate).toISOString()} as timestamp),
```
Returns:
```
            '  Raw query failed. Code: `22P03`. Message: `db error: ERROR: incorrect binary data format in bind parameter 1`',
```

I've exhausted my options here. How do I get this to work? All of these methods work without using variables so I'm really confused...

**Comment by leo-petrucci** [comment, marked as answer]
This works when using `$queryRawUnsafe()`, so I'm assuming it has something to do with the way variables are escaped?

Is it safe to escape variables myself and then use `queryRawUnsafe`?

I'm thinking this is most likely a bug, so I might open an issue in a bit.

**Reply by ryands17** [reply]
This looks like a bug. It would be great if you could create one [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=kind%2Fbug&template=bug_report.yml) to that we can look into this. As for the workaround, yes you can escape variables and use `queryRawUnsafe`.

**Reply by leo-petrucci** [reply]
Thank you for the reply, opened at #9197

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
